README.rst
.. image:: https://travis-ci.org/ant30/s3authbasic.svg?branch=master
  :target: https://travis-ci.org/ant30/s3authbasic

.. image:: https://coveralls.io/repos/ant30/s3authbasic/badge.png?branch=master
  :target: https://coveralls.io/r/ant30/s3authbasic?branch=master

.. contents::

===========
s3authbasic
===========


s3authbasic is a Pyramid application aimed to serve a static websites stored
in Amazon S3 protected by Auth Basic authentication. You need a IAM user with
its credentials and read access on the bucket which contains the static
files.

Amazon Settings
===============

I assumed that you have a protected bucket. To create the IAM user with the
correct permissions you need go to the IAM_ app in the AWS Console

.. _IAM: https://console.aws.amazon.com/s3/home

Click **Users** section and click in the **Creare New User** blue button.

Enter the username you want. It can't contains whitespaces. And click on
**create**

Click on **Download Credential** or write down the credentials info.

Now, we need to add the policy to give read access permissions to the user.

Click in the created user and go to the **permissions** tab below.

Click on the **Attach User Policy**.

Click on **Custom Policy** and in the **Select** button.

Give a policy name, like **s3-reader-policy**.

Copy the follow block in the **Policy Document** box. Please,
replace **protected-html** with your bucket name.

.. code-block:: javascript

   {
       "Statement": [
        {
          "Effect": "Allow",
          "Action": [
            "s3:Get*",
            "s3:ListBucket"
          ],
          "Resource": [
            "arn:aws:s3:::protected-html",
            "arn:aws:s3:::protected-html/*"
          ]
        }
      ]
   }


Application deployment
======================

I suggest you to use a virtualenv but I don't use that in this doc.

.. code-block:: bash

   pip install s3authbasic


You need a Pyramid ini settings file. You can find one example_ in the
github reposity in the config-templates directory.

.. _example: https://github.com/ant30/s3authbasic/blob/master/s3authbasic/config-templates/development.ini


Configuration
=============

You can config the app by editing the settings file or by using envionment
variables.

If you want to use the file, then the AWS and the user settings are
documented in the file.

The AWS environment variables are like this:

.. code-block:: bash

   export AWS_BUCKET_NAME='bucketname'
   export AWS_SECRET_ACCESS_KEY='12312sdf32'
   export AES_ACCESS_KEY_ID='123123123aaa'

To define users to access the app, the variable should starts with **USER_**.
The user name is the word after the underscore character and shoud be defined
in the correct capitalization. This is, if you want a user called **admin**,
you should create the environment variable USER_admin.

You can setup many users as you need.

The password should be hashed by sha256.

You can get the hash of your password by this command in linux:

.. code-block:: bash

   echo -n thepassword | sha256sum

The environment variable should look like this:

.. code-block:: bash

   export USER_admin='123123123123123123123'

**NOTE:** In 0.1.3 version we added encoding utf-8 option to password, so if
you want to use non ascii characters, the hash can vary. You can use something
like this in a terminal. You need the hashlib package from pypi.

.. code-block:: bash

   from hashlib import sha256
   sha256('your password!!'.encode('utf-8')).hexdigest()



Start the application
=====================

You should use the follow command if you have customized the ini file:

.. code-block:: bash

   pserve development.ini


Otherwise, you can use this command:

.. code-block:: bash

   run-s3authbasic

If you are using the default ini file, the application should be
available in the 6543 port. If you have deployed the app in your
own system, you can access it through http://127.0.0.1:6543/


CHANGES.txt
0.1.3
-----

- Fix UTF8 passwords problems

0.1.2
-----

- Fix path with whitespaces
- Fix settings relative location and the default package location detection

0.1.1
-----

- Fix WebTest error
- Add files needed by heroku

0.1
---

- The first published version

0.0
---

-  Initial version


s3authbasic/tests.py
import unittest

from pyramid import testing


class ViewTests(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def test_my_view(self):
        from .views import my_view
        request = testing.DummyRequest()
        info = my_view(request)
        self.assertEqual(info['project'], 's3authbasic')


s3authbasic/tests/test_views.py
from s3authbasic.testing import BaseAppTest, AUTH_ENVIRON


class ViewsTests(BaseAppTest):

    def test_validpath(self):
        for (path, expect) in (
            ('/', 'home'),
            ('/index.html', 'home'),
            ('/level1', 'level 1'),
            ('/level1/', 'level 1'),
            ('/level1/index.html', 'level 1'),
            ('/level1/other.html', 'other'),
            ('/level1/level2', 'level 2'),
            ('/level1/level2/index.html', 'level 2'),
            ('/blue%20child', 'whitespace'),
            ('/blue%20child/index.html', 'whitespace'),
        ):
            result = self.testapp.get(path, extra_environ=AUTH_ENVIRON,
                                      status=200)
            self.assertTrue(expect in result.body)

    def test_not_validpath(self):
        for path in (
            '/other.html',
            '/index',
            '/level1/index',
            '/level3/',
        ):
            self.testapp.get(path, extra_environ=AUTH_ENVIRON,
                             status=404)


class MimeTypeTests(BaseAppTest):

    def test_mimetypes(self):
        for (path, mimetype) in (
            ('/', 'text/html'),
            ('/index.html', 'text/html'),
            ('/example.jpeg', 'image/jpeg'),
        ):
            result = self.testapp.get(path, extra_environ=AUTH_ENVIRON)
            self.assertEqual(result.content_type, mimetype)


s3authbasic/tests/test_runapp.py
import os
import unittest

from s3authbasic.runapp import main
from s3authbasic.testing import unset_environment

environ = {
    'USER_admin': '8c6976e5b5410415bde908bd4dee15dfb16',
    'AWS_BUCKET_NAME': 'bucket_example',
    'AWS_ACCESS_KEY_ID': '123123',
    'AWS_SECRET_ACCESS_KEY': '123123123',
}


def mock_serve(app, host='0.0.0.0', port='6543', url_scheme='http'):
    return {
        'app': app,
        'host': host,
        'port': port,
        'url_scheme': url_scheme,
    }


class RunAppTest(unittest.TestCase):

    def setUp(self):
        os.environ.update(environ)

    def test_default_run(self):
        app = main(serve=mock_serve)
        self.assertTrue(app['host'] == '0.0.0.0')

    def test_change_port(self):
        os.environ.update({
            'PORT': '9999'
        })
        app = main(serve=mock_serve)
        self.assertTrue(app['port'] == 9999)

    def test_change_scheme(self):
        app = main(serve=mock_serve)
        self.assertTrue(app['url_scheme'] == 'https')
        os.environ.update({
            'SCHEME': 'http'
        })
        app = main(serve=mock_serve)
        self.assertTrue(app['url_scheme'] == 'http')

    def test_change_settings(self):
        os.environ.update({
            'SETTINGS': 's3authbasic/config-templates/production.ini'
        })
        app = main(serve=mock_serve)
        self.assertTrue(app['url_scheme'] == 'http')

    def tearDown(self):
        unset_environment()


s3authbasic/__init__.py
import os
from hashlib import sha256
import re

from pyramid.authentication import BasicAuthAuthenticationPolicy
from pyramid.authorization import ACLAuthorizationPolicy
from pyramid.config import Configurator
from pyramid.exceptions import ConfigurationError
from pyramid.security import Allow, Authenticated

from .s3controller import S3Controller, get_s3_controller


def check_credentials(username, password, request):
    credentials = request.registry.settings['credentials']
    passwordhash = credentials.get(username, None)

    passwordhashed = sha256(password.encode('utf-8')).hexdigest()
    if passwordhash is not None and passwordhashed == passwordhash:
        return [username]


def read_settings(settings, key, default=None, required=False):
    env_key = key.upper()
    value = None
    if env_key in os.environ:
        value = os.environ.get(env_key, default)
    else:
        value = settings.get(key, default)

    if value is None and required:
        raise ConfigurationError('The property {0} is required'.format(key))

    return value


def read_users(settings):
    credentials = {}
    key_prefix = 'user_'

    for key in settings.keys():
        if key.startswith(key_prefix):
            username = key.replace(key_prefix, '')
            password = settings.get(key)
            credentials[username] = password

    for key in os.environ.keys():
        if key.startswith(key_prefix.upper()):
            username = key.replace(key_prefix.upper(), '')
            password = os.environ.get(key)
            credentials[username] = password

    if not credentials:
        raise ConfigurationError("There isn't any user set, please use keys"
                                 " like user_admin = 123wdsdd123123 (encoded "
                                 "via sha256)")
    return credentials


class RootSite(object):
    __acl__ = [
        (Allow, Authenticated, 'view'),
    ]

    def __init__(self, request):
        pass


def main(global_config, **settings):
    """ This function returns a Pyramid WSGI application.
    """
    settings['credentials'] = read_users(settings)
    settings['realm'] = read_settings(settings, 'realm',
                                      default='Protected site')

    settings['aws_bucket_name'] = read_settings(
        settings, 'aws_bucket_name', required=True)

    settings['aws_secret_access_key'] = read_settings(
        settings, 'aws_secret_access_key', required=True)

    settings['aws_access_key_id'] = read_settings(
        settings, 'aws_access_key_id', required=True)

    settings['s3'] = S3Controller(settings)

    authn_policy = BasicAuthAuthenticationPolicy(check_credentials,
                                                 realm=settings['realm'])
    authz_policy = ACLAuthorizationPolicy()

    config = Configurator(settings=settings,
                          root_factory=RootSite)
    config.set_authentication_policy(authn_policy)
    config.set_authorization_policy(authz_policy)
    config.add_route('site', pattern='{path:.*}')
    config.add_request_method(get_s3_controller, name='s3', reify=True)

    config.scan(ignore=[re.compile('.*test(s|ing).*').search])
    return config.make_wsgi_app()


s3authbasic/s3controller.py
from boto.s3.connection import S3Connection


class S3Controller:

    def __init__(self, settings):
        self.key_id = settings['aws_access_key_id']
        self.secret = settings['aws_secret_access_key']
        self.bucket_name = settings['aws_bucket_name']

        s3 = settings.get('S3Wrapper', S3Connection)

        self.s3 = s3(self.key_id, self.secret)
        self.bucket = self.s3.get_bucket(self.bucket_name, validate=False)

    def get_file(self, path):
        if path == '/':
            path = 'index.html'
        s3file = self.bucket.lookup(path)
        if s3file is None:
            if not path.endswith('/'):
                path += '/'
            s3file = self.bucket.lookup(path + 'index.html')
        return s3file


def get_s3_controller(request):
    return request.registry.settings['s3']


s3authbasic/testing.py
from mimetypes import MimeTypes
import os
from os import path
import urllib

import unittest
from webtest import TestApp

from . import main


BASEPATH = path.abspath(path.dirname(__file__))
FIXTURES = path.join(BASEPATH, 'tests/fixtures')

AUTH_ENVIRON = {
    'HTTP_AUTHORIZATION': 'Basic YWRtaW46YWRtaW4='
}


class S3FileMockup(object):

    def __init__(self, filepath, block_size=1024):
        mime = MimeTypes()
        self.block_size = block_size
        url = urllib.pathname2url(filepath)
        self.content_type = mime.guess_type(url)[0]
        self.file = open(filepath)

    def __iter__(self):
        return self

    def next(self):
        val = self.file.read(self.block_size)
        if not val:
            raise StopIteration
        return val

    #  py3
    __next__ = next

    def close(self):
        self.file.close()


class S3BucketMockup:

    def __init__(self, bucket_name):
        self.bucket_name = bucket_name
        self.abspath = path.join(FIXTURES, bucket_name)

    def lookup(self, filename):
        if filename.startswith('/'):
            filename = filename[1:]

        file = path.join(self.abspath, filename)
        if path.isfile(file):
            return S3FileMockup(file)
        return None


class S3ConnectionMockup:

    def __init__(self, key, secret):
        self.key_id = key
        self.key_secret = secret

    def get_bucket(self, bucket_name, validate=True):
        return S3BucketMockup(bucket_name)


def unset_environment():
    for item in [
        'USER_admin',
        'AWS_BUCKET_NAME',
        'AWS_ACCESS_KEY_ID',
        'AWS_SECRET_ACCESS_KEY',
    ]:
        if item in os.environ:
            del os.environ[item]


class BaseAppTest(unittest.TestCase):

    def setUp(self):
        unset_environment()
        settings = {
            'user_admin': '8c6976e5b5410415bde908bd4dee15dfb16'
                          '7a9c873fc4bb8a81f6f2ab448a918',
            'aws_bucket_name': 'bucket_example',
            'aws_access_key_id': '123123',
            'aws_secret_access_key': '123123123',
            'S3Wrapper': S3ConnectionMockup,
        }

        app = main({}, **settings)
        self.testapp = TestApp(app)

    def tearDown(self):
        unset_environment()
        self.testapp.reset()
